Graphql Schema Update

aqeelmozumder opened the pull request:
Added Compnay enum

Have added changes for GenerateScheduleInput:

- Added Term
- Added Course

Have added changes for User:

- Added displayName
- Added hasPeng 

Have added some changes required by front end:

High Priority Changes:

- Added mutation for editing schedule
- Added Section Number
- Added Non-teaching term
- Added Teaching capacity
- Added Has Relief
- Added Relief Explanation
- Added Has topic
- Added Topic Description
- Added Display Name

Low Priority Changes:

- Added Course Title
- Added Query to grab all course preferences
- Added Query for all users

keithradford commented on graphql/schema.graphql:
```diff
@@ -1,57 +1,95 @@
-  """Logout the currently logged in user"""
+  """
+  Logout the currently logged in user
+  """
```
Kinda unrelated, but do we actually need a `logout` mutation? I thought all the frontend would have to do is remove the token from session storage, doesn't need backend logic

aomi replied:
Since it's cookie-based auth you need to tell the backend to delete the session from the backend session storage.

LacroixJ replied:
We use it to invalidate the session on backend which seems to be best practice from a security standpoint. It's only 1 line of code.

keithradford replied:
ok. We just won't implement any logic cause it's kinda different on or end

aqeelmozumder replied:
I think for the dev, we might need it as we destroy the session in the context or we could just re run the program every time.

Adamk8 commented on graphql/schema.graphql:
```diff
@@ -85,69 +128,121 @@ type CoursePreference {
 input GenerateScheduleInput {
   year: Int!
+  term: Term!
+  courses: CourseID!
```
Should this be [CourseID!] because we will send multiple courses?

keithradford replied:
courses should be an array, no?

keithradford commented on graphql/schema.graphql:
```diff
@@ -1,57 +1,95 @@
+  """
+  Edit schedule
+  """
+  editSchedule(year: Int!): Response!
```
First of all, wouldn't you need to supply data to edit the schedule instead of just yeaR=r?

And if this is the solution to editable schedules on the frontend, I've been thinking maybe it makes more sense to edit the courses meeting times directly, instead of the whole schedule?
My understanding and correct me if I'm wrong is that a schedule object points to a bunch of different courses that make up the schedule. So if you edit the course that's being pointed to, it will still be reflected on the calendar but you don't need to edit the whole schedule.

Sorry for the long comment lmk if that makes sense

aomi replied:
Your point makes sense 👍

LacroixJ replied:
TBH edit schedule is unrealistic to have working for the next demo, I think we should drop this entirely for now.

aqeelmozumder replied:
Makes sense,  its more of editing the courses instead of editing the schedule. 

If we are editing courses, then it I think the input would be changing the course meeting times, section number, etc, so basically courseSection? instead of year

@Adamk8, would front end have any concerns? because I am not sure if editing the courses in a schedule would automatically update the calendar?

keithradford commented on graphql/schema.graphql:
```diff
@@ -85,69 +128,121 @@ type CoursePreference {
 
 type CourseID {
-  """Course subject, e.g. SENG, CSC"""
+  """
```
based on recent meetings, do we need to add `section: String!` to this (i.e. A01, A02)

keithradford replied:
https://github.com/SENG499-S22-Company3/shared/pull/14#discussion_r914034562

keithradford commented on graphql/schema.graphql:
```diff
@@ -85,69 +128,121 @@ type CoursePreference {
+  """
+  Section number for courses, eg: A01, A02
+  """
+  sectionNumber: String
```
i take back my last comment, but does this make sense to be in courseID?

aomi commented on graphql/schema.graphql:
```diff
@@ -163,32 +258,62 @@ type CreateUserMutationResult {
+  """
+  Auth token used for future requests
+  """
   token: String!
```
Can we remove this or add the `@deprecated` directive here?

aqeelmozumder commented:
**Commit 2**

- Created a new input for editSchedule, EditCourseInput
- Created a company mutation to choose a company's algorithm
- Added CourseInput for  courses in GenerateScheduleInput
- Added @deprecated for token

Adamk8 commented on graphql/schema.graphql:
```diff
@@ -85,69 +132,152 @@ type CoursePreference {
+  endTime: Date!
+}
+
+input CourseInput {
```
CourseInput needs to have `sections: Int!` as well

aomi commented on graphql/schema.graphql:
```diff
@@ -1,57 +1,99 @@
+  """
+  Plug and Play Compnay 3's and 4's Algorithms
+  """
+  company(input: CompanyInput): Response!
```
I thought we wanted `CompanyInput` to be part of `generateSchedule`?
I'm fine with either but just want to make sure.

Adamk8 replied:
Either way would work, but I think it makes more sense to have 1 mutation instead of 2

aqeelmozumder replied:
I will make it a single mutation then

aqeelmozumder commented:
Commit 3 and 4:

Added single mutation within GenerateScheduleInput to choose company 3's and 4's Algorithms
Added sections:Int! in CourseInput


Not sure about the linting issue

keithradford commented on graphql/schema.graphql:
```diff
@@ -146,7 +146,39 @@ input CoursePreferenceInput {
+  """
+  Number of Sections for a course
+  """
+  sections: Int!
```
```suggestion
  section: Int!
```
naming nit

keithradford commented on graphql/schema.graphql:
```diff
@@ -87,9 +87,9 @@ type Mutation {
+  Edit schedule and Plug and Play Compnay 3's and 4's Algorithms
   """
-  editSchedule(year: Int!): Response!
+  editSchedule(input: EditCourseInput): Response!
```
```suggestion
  editCourse(input: EditCourseInput): Response!
```
since you can edit capacity and stuff too, I think `editCourse` probably makes more sense

aomi replied:
Adding some context for discussion here.
<img width="868" alt="image" src="https://user-images.githubusercontent.com/1757744/177622498-114c940b-7c4d-420d-b57e-46e4c1bc62e5.png">

keithradford commented on graphql/schema.graphql:
```diff
@@ -146,7 +146,39 @@ input CoursePreferenceInput {
+  algorithm2: Company!
+}
+
+input EditCourseInput {
```
Shouldn't there be some attribute to identify a course? Looks like just `sectionNumber` which I assume is A01, etc, but how do we know what course to edit in the mutation?

keithradford replied:
Could add the `CourseInput` type perhaps and then remove `sectioNumber` since it's a part of course input

aqeelmozumder replied:
I dont think we should remove sectionNumber. As sectionNumber is not a part of CourseInput but section is.

keithradford replied:
Gotcha

aqeelmozumder commented:
**Commit 6,7,8:**

`sections: Int!` now go by `section: Int!`
Mutation `editSchedule` renamed as `editCourses`
Added `CourseInput` in `EditCourseInput` to identify which course needs to be edited
